adr/ADR-15.md

```diff
@@ -0,0 +1,235 @@
+	Description     string        `json:"description,omitempty"`
+	Durable         string        `json:"durable_name,omitempty"`
+	FilterSubject   string        `json:"filter_subject,omitempty"`
+	FlowControl     bool          `json:"flow_control,omitempty"`
+	Heartbeat       time.Duration `json:"idle_heartbeat,omitempty"`
+	MaxAckPending   int           `json:"max_ack_pending,omitempty"`
+	MaxDeliver      int           `json:"max_deliver,omitempty"`
+	MaxWaiting      int           `json:"max_waiting,omitempty"`	// server returns error if provided if a DeliverSubject is provided
+	OptStartSeq     uint64        `json:"opt_start_seq,omitempty"`	// server returns error if provided when Deliver Policy is not by_start_sequence
+	OptStartTime    *time.Time    `json:"opt_start_time,omitempty"`	// server returns error if provided when Deliver Policy is not by_start_time
+	RateLimit       uint64        `json:"rate_limit_bps,omitempty"` // Bits per sec
+	ReplayPolicy    ReplayPolicy  `json:"replay_policy"`
+	SampleFrequency string        `json:"sample_freq,omitempty"`
+}
+```
+
+#### Stream name
+
+When no stream is specified, the `subject` to the subscribe API calls is required. An error will be returned if it is not the case.
+The library will use the subject provided as a way to find out which stream this subscription is for. A request is sent to the server on the `<JS prefix>.STREAM.NAMES` subject with a JSON content `{"subject":"<subject here>"}`. If the response (a list of stream names) is positive and contains a single entry, then the library will use this stream name, otherwise an error indicating that there is no matching stream name should be returned.
+
+If the stream name is specified, then the `subject` is possibly not needed.
+
+#### Consumer name
+
+When a consumer name is specified, it indicates the the library that the intent is to bind to an existing JS Consumer. The library will lookup the consumer from the server and get a `ConsumerInfo`. If the lookup fails, it is considered an error, unless the subscription is for a "pull subscriber", in which case the library still proceeds with the JS subscription.
+
+#### Queue name
+
+If the user attempts to create a queue subscription, a consumer name or durable needs to be specified. The common pattern error that we noticed user doing was this:
+
+```go
+member1, _ := js.QueueSubscribe("foo", "bar")
+member2, _ := js.QueueSubscribe("foo", "bar")
+```
+And then report that both members were receiving the same message.
+
+This is because each subscription call would create their own ephemeral JetStream consumer. We could break the API and force the user to specify a consumer name. Instead, for the Go client we have taken the approach to describe that if no consumer name is provided (in Go with `nats.Bind(stream, consumer)`) nor a durable (in Go with `nats.Durable(durableName)`) then the library would use the queue name as the durable name.
```

Comment:
With the new `deliver_group` field in the consumer config, the queue name should be used as the deliver group. is a durable name even required when a deliver group is specified, or can ephemerals participate in a deliver group? 
If ephemerals do not participate in a user group, I suppose there is no reason why the durable name cannot be the same as the deliver group name, but I would check with the server to see if this is acceptable.
Also, should we retire the queue concept for deliver group concept? I think queue group is actually a confusing name to begin with because queue really means something else. I know, naming is hard.

Reply:
The confusion you have is that an ephemeral consumer is not a durable, but it is still a consumer and it has a name. (but happen - as of now - to be selected by the user when it is created). This is why I describe above that I can't change "consumer name" to "durable" everywhere because it is not correct.
(for information: a durable is replicated across server nodes, while an ephemeral is always "R1", which makes it faster).
As explained earlier, one *could* create a queue group (call many instances of js.QueueSubscribe() with the same queue name) for an ephemeral JS consumer, using `nats.Bind(stream, consumer)` option, in that case without `nats.Durable()` option. That is, an admin could create an ephemeral (with a given deliver group) and give that name to application owners. Again, unlikely, but possible.